packages/router/src/grpc/offscreen-client.ts

```diff
@@ -61,29 +62,35 @@ const buildActions = (
   fullViewingKey: FullViewingKey,
   cancel: PromiseLike<never>,
 ): Promise<Action>[] => {
-  const active = activateOffscreen();
-  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  const activation = activateOffscreen();
+
+  // this is a lot of binary -> base64 serialization, so just do it once and reuse
+  const partialRequest = {
+    transactionPlan: transactionPlan.toJson() as Jsonified<TransactionPlan>,
+    witness: witness.toJson() as Jsonified<WitnessData>,
+    fullViewingKey: fullViewingKey.toJson() as Jsonified<FullViewingKey>,
+  };
+
   const buildTasks = transactionPlan.actions.map(async (_, actionPlanIndex) => {
-    await active;
-    const buildRes = await sendOffscreenMessage<ActionBuildMessage>({
+    const buildReq: InternalRequest<ActionBuildMessage> = {
       type: 'BUILD_ACTION',
       request: {
-        transactionPlan: transactionPlan.toJson() as Jsonified<TransactionPlan>,
-        witness: witness.toJson() as Jsonified<WitnessData>,
-        fullViewingKey: fullViewingKey.toJson() as Jsonified<FullViewingKey>,
+        ...partialRequest,
         actionPlanIndex,
-      } satisfies ActionBuildRequest,
-    });
-    if ('error' in buildRes) throw ConnectError.from(buildRes.error);
```

Comment:
`throw ConnectError.from(buildRes.error);` was technically wrong. needed to use `errorFromJson`